db/inlineskiplist.h

```diff
@@ -529,63 +587,123 @@ InlineSkipList<Comparator>::AllocateNode(size_t key_size, int height) {
+  hint->last_height = 0;
+  hint->prev_height = 1;
+  for (int i = 0; i < kMaxHeight_; i++) {
+    hint->prev[i] = head_;
   }
+  return hint;
+}
 
-  // Our data structure does not allow duplicate insertion
-  assert(prev_[0]->Next(0) == nullptr || !Equal(key, prev_[0]->Next(0)->Key()));
-
+template <class Comparator>
+typename InlineSkipList<Comparator>::Node*
+InlineSkipList<Comparator>::GetNodeForInsert(const char* key, int* height) {
   // Find the Node that we placed before the key in AllocateKey
   Node* x = reinterpret_cast<Node*>(const_cast<char*>(key)) - 1;
-  int height = x->UnstashHeight();
-  assert(height >= 1 && height <= kMaxHeight_);
-
-  if (height > GetMaxHeight()) {
-    for (int i = GetMaxHeight(); i < height; i++) {
-      prev_[i] = head_;
-    }
+  assert(height != nullptr);
+  *height = x->UnstashHeight();
+  assert(*height >= 1 && *height <= kMaxHeight_);
 
+  if (*height > GetMaxHeight()) {
     // It is ok to mutate max_height_ without any synchronization
     // with concurrent readers.  A concurrent reader that observes
     // the new value of max_height_ will see either the old value of
     // new level pointers from head_ (nullptr), or a new value set in
     // the loop below.  In the former case the reader will
     // immediately drop to the next level since nullptr sorts after all
     // keys.  In the latter case the reader will use the new node.
-    max_height_.store(height, std::memory_order_relaxed);
+    max_height_.store(*height, std::memory_order_relaxed);
+  }
+
+  return x;
+}
+
+template <class Comparator>
+void InlineSkipList<Comparator>::Insert(const char* key) {
+  int height = 0;
+  Node* x = GetNodeForInsert(key, &height);
+
+  // InsertConcurrently often can't maintain the prev_ invariants, so
+  // it just sets prev_height_ to zero, letting us know that we should
+  // ignore it.  A relaxed load suffices here because write thread
+  // synchronization separates Insert calls from InsertConcurrently calls.
+
+  // Outside of this method prev_[1..max_height_] is the predecessor
+  // of prev_[0], and prev_height_ refers to prev_[0].  Inside Insert
+  // prev_[0..max_height - 1] is the predecessor of key.  Switch from
+  // the external state to the internal
+
+  bool hint_valid = hint_valid_.exchange(true, std::memory_order_relaxed);
+  if (!hint_valid || !KeyIsAfterNode(key, hint_->Last()) ||
+      KeyIsAfterNode(key, hint_->Last()->NoBarrier_Next(0))) {
+    // TODO(opt): we could use a NoBarrier predecessor search as an
+    // optimization for architectures where memory_order_acquire needs
+    // a synchronization instruction.  Doesn't matter on x86
+    FindLessThan(key, hint_->prev);
+    hint_->last_height = 0;
   }
 
-  for (int i = 0; i < height; i++) {
-    // NoBarrier_SetNext() suffices since we will add a barrier when
-    // we publish a pointer to "x" in prev[i].
-    x->NoBarrier_SetNext(i, prev_[i]->NoBarrier_Next(i));
-    prev_[i]->SetNext(i, x);
```

Comment:
PushLast() captures line 567-569 and line 581-585 in the existing code and they are doing exactly the same thing.